packages/types/Result.ts

```diff
@@ -13,40 +13,77 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see https://www.gnu.org/licenses/.
  */
+import createError from 'http-errors';
+import { assertExhaustive } from './assertExhaustive';
 
 type ResultBase = {
   readonly _tag: 'Result';
 };
 
+export enum ErrorResultKind {
```

Comment:
I'm not sure if I'd like the idea of replacing a HTTP error code with a text based error category that maps 1-1 to a HTTP error code. All this does is prevent the business logic code having to know HTTP error code numbers, but it still has to figure out what the type of failure is in HTTP terms, so HTTP specific concerns are still bleeding into the business logic layer - we have to assign errors in there with might not make so much sense if we were publishing the result to a queue, for example.

I think the business layer should return errors that make sense within their own context, and then it should be up to the route layer to determine what sort of HTTP response this becomes - in some cases 'success' results in the business layer may become errors in the HTTP layer (null / undefined -> 404 comes to mind) or vice versa

I'm also not sure it makes sense to create a definitive set of error categories at the top level that all the business layer should use. Each function within the service should be able to define their own error schemes that make sense in their context?

Happy to chat about this more if you like

Reply:
Yep I was also considering if this made more sense than just the numbers. Thanks for the feedback. I agree with you, so I'll be moving to a more "each service/layer defines it's own errors" approach.